text/0000-ptr-to-field.md

```diff
@@ -0,0 +1,426 @@
+
+This RFC introduces the idea of field types, and build projections on top of field types.
+
+## Terminology, types, and traits
+
+### Field Types
+
+A field type is a compiler generated type that represents a field. For example,
+
+```rust
+struct Person {
+    name: String,
+    age: u32,
+}
+```
+
+A field type for the `name` field of `Person` would be `Person.name`.
+
+<sub>Note about syntax: the syntax `Type.field` is going to be used as a placeholder syntax for the rest of this RFC, but this is *not* the final syntax for field types, that decision will be made after this RFC gets accepted but before it gets stabilized.</sub>
+
+Field types serve one important purpose, because they are types, they can implement traits, and thus be generalized over. This allows for powerful compile time guarantees that field types are being used correctly.
+
+### `trait Field` and `type FieldDescriptor`
+
+In order to generalize over field types, we have the `trait Field` 
+
+```rust
+trait Field {
+    /// The type of the type that the field comes from
+    type Parent: ?Sized;
+
+    /// The type of the field itself
+    type Type: ?Sized;
+
+    const FIELD_DESCRIPTOR: FieldDescriptor<Self>;
+}
+
+struct FieldDescriptor<F: Field + ?Sized> { .. }
+```
+
+`FieldDescriptor` is an opaque type that will store some metadata about how to convert from a `*const Field::Parent` to a `*const Field::Type`. There is no way to safely construct `FieldDescriptor` from user code on Stable Rust until Rust gets a defined stable type layout for `repr(Rust)` types.
+
+The `Field` trait will allow generalizing over field types, and thus allow other apis to be created, for example...
+
+### `*const T`/`*mut T` methods
+
+We will add the following methods to raw pointers
+
+```rust
+// details about why we need both and what they do exactly in Reference-level explanation
+
+impl<T: ?Sized> *const T {
+    pub unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *const F::Type;
+    
+    pub fn wrapping_project<F: Field<Parent = T>>(self, field: F) -> *const F::Type;
+
+    pub unsafe fn inverse_project_unchecked<F: Field<Type = T>>(self, field: F) -> *const F::Parent;
+    
+    pub fn inverse_wrapping_project<F: Field<Type = T>>(self, field: F) -> *const F::Parent;
+}
+
+impl<T: ?Sized> *mut T {
+    pub unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *mut F::Type;
+    
+    pub fn wrapping_project<F: Field<Parent = T>>(self, field: F) -> *mut F::Type;
+    
+    pub unsafe fn inverse_project_unchecked<F: Field<Type = T>>(self, field: F) -> *mut F::Parent;
+    
+    pub fn inverse_wrapping_project<F: Field<Type = T>>(self, field: F) -> *mut F::Parent;
+}
+```
+
+These will allowing projections through raw pointers without dereferencing the raw pointer. This is useful for building projections through other abstractions like smart pointers (`Rc<T>`, `Pin<&T>`)!
+
+This is the extent of the core api of this RFC.
```

Comment:
Is this line necessary anymore?

Reply:
No, I seemed to have missed that in my last read through last night.